src/trace_ng/log.zig

```diff
@@ -0,0 +1,1048 @@
+            self.logWithFields(.debug, message, fields);
+        }
+
+        pub fn debugfWithFields(self: *Self, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            self.logfWithFields(.debug, fmt, args, fields);
+        }
+
+        pub fn log(self: *Self, level: Level, message: []const u8) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |*impl| impl.*.log(level, message),
+            }
+        }
+
+        pub fn logf(self: *Self, level: Level, comptime fmt: []const u8, args: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logf(level, fmt, args),
+            }
+        }
+
+        pub fn logWithFields(self: *Self, level: Level, message: []const u8, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logWithFields(level, message, fields),
+            }
+        }
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logfWithFields(level, fmt, args, fields),
+            }
+        }
+    };
+}
+
+pub const Logger = ScoppedLogger(null);
+
+/// An instance of `ScopedLogger` that logs to the standard err.
+pub fn StandardErrLogger(comptime scope: ?[]const u8) type {
+    return struct {
+        const Self = @This();
+        max_level: Level,
+        exit_sig: std.atomic.Value(bool),
+        allocator: Allocator,
+        log_allocator: Allocator,
+        log_allocator_state: *RecycleFBA(.{}),
+        max_buffer: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) !*Self {
+            const recycle_fba = try config.allocator.create(RecycleFBA(.{}));
+            const max_buffer = config.max_buffer orelse return error.MaxBufferNotSet;
+            recycle_fba.* = try RecycleFBA(.{}).init(config.allocator, max_buffer);
+            const self = try config.allocator.create(Self);
+            self.* = .{
+                .allocator = config.allocator,
+                .log_allocator = recycle_fba.allocator(),
+                .log_allocator_state = recycle_fba,
+                .max_buffer = max_buffer,
+                .exit_sig = AtomicBool.init(false),
+                .max_level = config.max_level,
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = try std.Thread.spawn(.{}, run, .{self}),
+            };
+            return self;
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.channel.close();
+            if (self.handle) |*handle| {
+                std.time.sleep(std.time.ns_per_ms * 5);
```

Comment:
Not 100% sure if this is the best way to fix this, but the issue I found was that in the test case where the scope of the logger was switched (ie essentially having two loggers...the one used by the parent struct and the child struct) it often happens that the log from the child does not make it to the std err before the process dies. It seems the handle.join() sometimes does not work...but if I introduce a `std.debug.print` (or the sleep above) it consistently shows up.

Reply:
> the scope of the logger was switched (ie essentially having two loggers

This wouldn't be an issue if you use only a single instance of the underlying logger.